inputremapper/injection/injector.py

```diff
@@ -212,25 +247,20 @@ def _grab_devices(self) -> GroupSources:
         needed_devices = {}
 
         for mapping in self.preset:
-            for event in mapping.event_combination:
-                candidates: List[evdev.InputDevice] = [
-                    device
-                    for device in devices
-                    if event.code
-                    in device.capabilities(absinfo=False).get(event.type, [])
-                ]
-                if len(candidates) > 1:
-                    # there is more than on input device which can be used for this
-                    # event we choose only one determined by the ranking
-                    device = sorted(
-                        candidates, key=lambda d: ranking.index(classify(d))
-                    )[0]
-                elif len(candidates) == 1:
-                    device = candidates.pop()
-                else:
-                    logger.error("Could not find input for %s in %s", event, mapping)
-                    continue
-                needed_devices[device.path] = device
+            update_configs = {}
+            for input_config in mapping.input_combination:
+                if device := self._find_input_device(devices, input_config):
+                    needed_devices[device.path] = device
+                    if input_config.origin is None:
+                        update_configs[input_config] = input_config.modify(
+                            origin=get_device_hash(device)
+                        )
+
+            if update_configs:
+                combination = list(mapping.input_combination)
+                for old, new in update_configs.items():
+                    combination[combination.index(old)] = new
+                mapping.input_combination = combination
```

Comment:
Some comments would be nice here. Maybe also split it into some private methods to make it easier to read

Reply:
I just shuffled a bunch of functions around in the Injector. Hopefully it is now clearer what is going on.